[interp] Move from stack based to fully local var based design

monojenkins opened the pull request:
!! This PR is a copy of mono/mono#20663,  please do not edit or review it in this repo !!<br/>Do not automatically approve this PR:<br/><br/>* Consider how the changes affect configurations in this repo,<br/>* Check effects on files that are not mirrored,<br/>* Identify test cases that may be needed in this repo.<br/><br/>!! Merge the PR only after the original PR is merged !!<br/><hr/><br/>

SamMonoRT commented:
@naricc  ran the .NET Microbenchmarks with these changes and we see a good perf improvement. From naricc : 
Here is desktop interpreter: https://pvscmdupload.blob.core.windows.net/reports/interpreterchange/report_Daily_ca%3Dx64_cb%3Drefs-heads-sync-pr-20663-from-mono_co%3DUbuntu1804_cr%3Ddotnetruntime_cc%3DCompilationMode%3Dtiered-LLVM%3Dfalse-MonoAOT%3Dfalse-MonoInterpreter%3Dtrue-RunKind%3Dmicro_mono_bb%3Drefs-heads-master_2021-01-06.html

Here is Wasm:  https://pvscmdupload.blob.core.windows.net/reports/interpreterchange/report_Daily_ca%3Dx64_cb%3Drefs-heads-sync-pr-20663-from-mono_co%3DUbuntu1804_cr%3Ddotnetruntime_cc%3DCompilationMode%3Dwasm-RunKind%3Dmicro_bb%3Drefs-heads-master_2021-01-06.html

BrzVlad committed:
[interp] Move from stack based to fully local var based design

Instead of having instructions that push and pop from the stack, every instruction has explicit dreg and sregs.

While the purpose of this PR is mainly to make it easier to implement more advanced optimization in the future, it also has noticeable performance implications. The code is simplified because we no longer need to update and save the SP. However, the code for each instruction is bloated due to the addition of explicit source and destination offsets. This is counteracted by the reduction of the total number of instructions, since ldloc/stloc and moves become redundant and they are mostly optimized away, even in this implementation state. Here are the total number of executed opcodes as part of running the corlib test suite with the interp https://gist.github.com/BrzVlad/d62f504930b75cba4b870e6dbd947e90.
